File: dotnet/apps/database/domain/apps/rules/product/InventoryItemPartDisplayNameRule.cs
// <copyright file="InventoryItemPartDisplayNameDerivation.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Database.Domain
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Database.Derivations;
    using Meta;
    using Derivations.Rules;

    public class InventoryItemPartDisplayNameRule : Rule
    {
        public InventoryItemPartDisplayNameRule(MetaPopulation m) : base(m, new Guid("9bd924bb-3625-423b-b12d-b90d70f491af")) =>
            this.Patterns = new[]
            {
                m.InventoryItem.RolePattern(v => v.Part),
            };

        public override void Derive(ICycle cycle, IEnumerable<IObject> matches)
        {
            foreach (var @this in matches.Cast<InventoryItem>())
            {
                if (!@this.ExistPartDisplayName)
                {
                    @this.PartDisplayName = @this.Part.DisplayName;
                }
            }
        }
    }
}

File: typescript/modules/libs/apps-intranet/workspace/derivations/src/lib/rules/inventory-item-part-display-name.rule.ts
import { Composite, Dependency, RoleType } from '@allors/system/workspace/meta';
import { IRule } from '@allors/system/workspace/domain';
import { M } from '@allors/default/workspace/meta';
import { InventoryItem } from '@allors/default/workspace/domain';

export class InventoryItemPartDisplayNameRule implements IRule<InventoryItem> {
  objectType: Composite;
  roleType: RoleType;
  dependencies: Dependency[];

  constructor(m: M) {
    const { dependency: d } = m;

    this.objectType = m.InventoryItem;
    this.roleType = m.InventoryItem.PartDisplayName;

    this.dependencies = [d(m.InventoryItem, (v) => v.Part)];
  }

  derive(inventoryItem: InventoryItem) {
    return inventoryItem.Part?.Name ?? '';
  }
}
